nexus/src/app/sled.rs

```diff
@@ -231,4 +233,166 @@ impl super::Nexus {
+
+        let (.., authz_instance, db_instance) =
+            LookupPath::new(&opctx, &self.db_datastore)
+                .instance_id(instance_id)
+                .fetch_for(authz::Action::Read)
+                .await?;
+
+        let instance_nics = self
+            .db_datastore
+            .derive_guest_network_interface_info(&opctx, &authz_instance)
+            .await?;
+
+        // Lookup the physical host IP of the sled hosting this instance
+        let instance_sled_id = db_instance.runtime().sled_id;
+        let physical_host_ip: std::net::IpAddr =
+            (*self.sled_lookup(&self.opctx_alloc, &instance_sled_id).await?.ip)
+                .into();
+
+        let mut last_sled_id: Option<Uuid> = None;
+        loop {
+            let pagparams = DataPageParams {
+                marker: last_sled_id.as_ref(),
+                direction: dropshot::PaginationOrder::Ascending,
+                limit: std::num::NonZeroU32::new(10).unwrap(),
+            };
+
+            let sleds_page =
+                self.sleds_list(&self.opctx_alloc, &pagparams).await?;
+
+            for sled in &sleds_page {
+                // set_v2p not required for sled instance was allocated to, OPTE
+                // currently does that automatically
+                if sled.id() == instance_sled_id {
+                    continue;
+                }
+
+                let client = self.sled_client(&sled.id()).await?;
+
+                for nic in &instance_nics {
+                    // This function is idempotent: calling the set_v2p ioctl with
+                    // the same information is a no-op.
+                    client
+                        .set_v2p(&SetVirtualNetworkInterfaceHost {
+                            virtual_ip: nic.ip,
+                            virtual_mac: nic.mac.clone(),
+                            physical_host_ip,
+                            vni: nic.vni.clone(),
+                        })
+                        .await?;
+                }
+            }
+
+            if sleds_page.len() < 10 {
+                break;
+            }
+
+            if let Some(last) = sleds_page.last() {
+                last_sled_id = Some(last.id());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Ensure that the necessary v2p mappings for an instance are deleted
+    pub async fn delete_instance_v2p_mappings(
+        &self,
+        opctx: &OpContext,
+        instance_id: Uuid,
+    ) -> Result<(), Error> {
+        // For every sled that isn't the sled this instance was allocated to, delete
+        // the virtual to physical mapping for each of this instance's NICs. If
+        // there isn't a V2P mapping, del_v2p should be a no-op.
+        let (.., authz_instance, db_instance) =
+            LookupPath::new(&opctx, &self.db_datastore)
+                .instance_id(instance_id)
+                .fetch_for(authz::Action::Read)
+                .await?;
+
+        let instance_nics = self
```

Comment:
So, if we change the endoint in the sled agent to be something like `DELETE /v2p/{instance_id}`, we could remove this call entirely. We only need the instance ID to do that.

Reply:
I don't grok this - won't the OPTE ioctl require all fields (it doesn't know about instance or NIC id) to delete the mapping?

Reply:
Yeah I think you're right. I might have been conflating this with the simulated sled agent, where things are stored that way. It might be possible to send just the VNI + MAC, but I don't see that being a huge win.